src/omp/backend.f90

```diff
@@ -382,16 +386,59 @@ subroutine vecadd_omp(self, a, x, b, y)
 
     y%data = a*x%data + b*y%data ! fixme
 
+    ! error stop "OMP: VecAdd Not implemented"
+
   end subroutine vecadd_omp
 
   real(dp) function scalar_product_omp(self, x, y) result(s)
+
+    use mpi
+    
     implicit none
 
     class(omp_backend_t) :: self
     class(field_t), intent(in) :: x, y
+    integer, dimension(3) :: dims
+    integer :: dir
+    integer :: i, j, k
+    integer :: ierr
 
-    s = 0._dp
-
+    if ((x%dir /= y%dir) .or. (x%data_loc /= y%data_loc)) then
+      error stop "Called scalar product with incompatible fields"
+    end if
+    
+    dims = self%mesh%get_field_dims(x)
+    dir = x%dir
+
+    s = 0.0_dp
+    if (dir == DIR_C) then
+      !$omp parallel do reduction(+:s) collapse(3)
+      do k = 1, dims(3)
+        do j = 1, dims(2)
+          do i = 1, dims(1)
+            s = s + x%data(i, j, k) * y%data(i, j, k)
+          end do
```

Comment:
If you use the `scalar_product` operator with `DIR_C` arrays and think that it could make a difference in performance, there is a simple trick you can implement to enable vectorisation here as well. 

You just need to separate the last loop into two as
```
do i_outer = 1, (dims(1) - 1)/SZ + 1
  ! pragma with reduction on s and private i
  do i_inner = 1, SZ
    i = (i_outer - 1)*SZ + i_inner
    s = s + ...
  end do
end do
! a cleanup loop
do i = 1, mod(dims(1), SZ)
  s = s + ... !final index = i + dims(1)/SZ
end do
```

Reply:
Done - also reduces total amount of code.